src/extension/features/toolkit-reports/pages/outflow-over-time/utils.ts

```diff
@@ -0,0 +1,156 @@
+import type { Moment } from 'moment';
+
+declare module 'moment' {
+  interface Moment {
+    toUTCMoment(): Moment;
+  }
+}
+
+interface YNABTransaction {
+  outflow?: number;
+  inflow?: number;
+  accountId: string;
+  transferAccountId: string | null;
+  date: Moment;
+  /** YYYY-mm format */
+  month: string;
+}
+
+export function filterTransactions(
+  transactions: readonly YNABTransaction[],
+  filterOutAccounts: ReadonlySet<string>
+) {
+  return transactions
+    .filter((transaction) => {
+      return (
+        transaction.outflow !== undefined &&
+        transaction.outflow !== 0 &&
+        (transaction.inflow === undefined || transaction.inflow === 0)
+      );
+    })
+    .filter((transactionsWithoutInflow) => {
+      if (transactionsWithoutInflow.transferAccountId) {
+        // discard unless it transfers outside the selected account list
+        return filterOutAccounts.has(transactionsWithoutInflow.transferAccountId);
+      }
+      return !filterOutAccounts.has(transactionsWithoutInflow.accountId);
+    });
+}
+
+/**
+ * Filter the passed transactions to only include those within the specified date range.
+ * @param transactions The array of transactions to be filtered.
+ * @param startDate The start date of the desired date range.
+ * @param endDate The end date of the desired date range.
+ * @returns A new array containing only the transactions that fall within the specified date range.
+ */
+export function filterTransactionsByDate(
+  transactions: readonly YNABTransaction[],
+  startDate: Moment,
+  endDate: Moment
+) {
+  return transactions.filter((transaction) => {
+    return transaction.date >= startDate && transaction.date <= endDate;
+  });
+}
+
+type GroupedTransactions = Record<string, Record<string, YNABTransaction[]>>;
+
+export function groupTransactions(transactions: readonly YNABTransaction[]) {
+  const groupedByMonth = groupBy(transactions, 'month');
```

Comment:
I couldn't really get rid of `groupBy`, or at least it would be a non-trivial change. I ended up reimplementing it as a helper function.